File: AmicableNumbers/AmicableNumbers.py
import os, sys
import math
import time, math

def testForPrimeness(number):
    answer=True
    if number%2==0:
        return False
    else:
        for i in range(3,int(math.sqrt(number))):
            if number%i==0:
                return False
            else:
                i+=2
    return answer

def sumOfProperDivisors(number):
    iSum=0
    if testForPrimeness(number)==False:
        if number%2!=0:
            divisors = [x for x in range(number/2+1,3,-2)]
        else:
            divisors = [x for x in range(number/2+1,0,-1)]

	for y in divisors:
            if (number%y)==0:
                iSum+=y
	    else:
                continue
    return iSum
sum=0
for a in range(2,10000):
    b=sumOfProperDivisors(a)
    if b > a:
        if sumOfProperDivisors(b)==a:
            sum+=(a+b)

print sum

File: AmicableNumbers/AmicableNumbers.cpp
#include <iostream>
#include <list>
#include <cmath>
#include <math.h>

using namespace std;

bool testForPrimeness(int &numberRef);
int sumOfProperDivisors(int &numberRef);

int main(){
    int sum=0;
    for (int a=2; a < 10000;a++){
        int b=0;
        b=sumOfProperDivisors(a);
        if (b > a){
            if (sumOfProperDivisors(b)==a){
                sum+=(a+b);
            }
        }
    }
    cout << sum << endl;
}


bool testForPrimeness(int &numberRef){
	int i=0;
	if (numberRef%2==0 && numberRef !=2){
		return false;
	}
	else{
		for (i = 3; i < sqrt(numberRef)+1;){
			if (numberRef%i==0){
                return false;
			}
			else{
			    i+=2;
			}
		}
	}
	return true;
}


int sumOfProperDivisors(int &numberRef){
    if (numberRef==1){
	return 0;
	}
	int iSum=0;
	
	if (testForPrimeness(numberRef)==false){
	    int r,f,step;
	    r = floor(sqrt(numberRef));
	    if (r*r==numberRef){
		iSum+=1+r; r= r-1;
	    }
	    else{
		iSum=1;
	    }
	    if (numberRef%2!=0){
		f=3; step=2;
	    }
	    else{
		f=2;step=1;
	    }
	    while(f<r){
		if (numberRef%f==0){
		    iSum+=f+numberRef/f;
		}
		f=f+step;
	    }
	}
    return iSum;
}
